### PR TITLE
#169257416 Delete comment posted within the comment thread 

### PR DESCRIPTION
**Pull request by gildniy**
#### What does this PR do?
Deletes comment posted within the comment thread 
#### Description of Task to be completed?
- Create a Comment model and comments table with following fields: 
`comment`, `userId`, `requestId`, `isVisible`
- Make table 1:N relations between `comments` and `requests` table and `comments` and `users` table
- Create `deleteComment` method in the `Comment` controller file and `deleteCommentById` in `CommentService`
#### How should this be manually tested?
- clone the repo to the local machine
- create a `trip` and create a comment on that trip, and get the `id` of the comment from the response after creation.
- in postman with method `PATCH` on `/api/v1/comment/:commentId/delete` where `:commentId` is the `id` gotten from previous step, click `send` button
In the postman body response is displayed the delete comment object.
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#169257416](https://www.pivotaltracker.com/story/show/169257416)
#### Screenshots (if appropriate)
1. Create trip
image

2.  Create a comment
image

3. Comments displaying on trip request
image

4. Deleting a comment
image

#### Questions:
N/A

### CODE REVIEW

File: src/routes/api/comments.route.js [hunk 1]
```diff
@@ -34,6 +36,43 @@ const router = express.Router();
...
+ *                     type: boolean
+ *     responses:
+ *       200:
+ *         description: Successfully deleted
+ *       404:
+ *         description: Comment not found
+ *       403:
+ *         description: Unauthorized user
+ */
+router.patch('/comment/:commentId/delete', verifyUser, catchErrors(deleteComment));
```
Comment (RaymondMwaura) [inline src/routes/api/comments.route.js]: I think you should rename this route to:
router.delete('/comment/:commentId', verifyUser, catchErrors(deleteComment));

That is, use the 'router.delete' function since it is a deletion endpoint.

**Review by RaymondMwaura — CHANGES_REQUESTED** [review]
The functionality works as described. Good job. I think you should rename the route to a more conventional format  as I have mentioned on a comment on your file: ' src/routes/api/comments.route.js'

File: src/routes/api/comments.route.js (hunk 1 above)
Comment (gildniy) [reply src/routes/api/comments.route.js]: > I think you should rename this route to:
> router.delete('/comment/:commentId', verifyUser, catchErrors(deleteComment));
> 
> That is, use the 'router.delete' function since it is a deletion endpoint.

Thank you @RaymondMwaura for the review, but as described in the PT story, by deleting a comment it should not be removed from the database, but is hidden from being shown. That's why its `isVisible` status has to be changed from default `true` to `false` with a `PATCH` method and the comment.

**Review by RaymondMwaura — DISMISSED** [review]
I think you have implemented this feature well. Great job!

**Comment by RaymondMwaura** [comment]
> The functionality works as described. Good job. I think you should rename the route to a more conventional format as I have mentioned on a comment on your file: ' src/routes/api/comments.route.js'

Your explanation that the story requires that the deleted comments remain in the database now explains why the endpoint is a PATCH and not a DELETE endpoint.

File: src/app.js [hunk 2]
```diff
@@ -36,9 +36,10 @@ app.use(cors({
     if (!origin) return callback(null, true);
     if (allowedOrigins.indexOf(origin) === -1) {
       const msg = 'The CORS policy for this site does not '
-                + 'allow access from the specified Origin.';
+        + 'allow access from the specified Origin.';
       return callback(new Error(msg), false);
-    } return callback(null, true);
+    }
+    return callback(null, true);
```
Comment (collinewait) [inline src/app.js]: Are changes in this file necessary?

File: src/controllers/auth.controller.js [hunk 3]
```diff
@@ -65,7 +65,7 @@ class AuthController {
         isVerified: true
       }
     })
-    // eslint-disable-next-line no-unused-vars
+      // eslint-disable-next-line no-unused-vars
```
Comment (collinewait) [inline src/controllers/auth.controller.js]: Are changes in this file necessary?

File: src/controllers/request.controller.js [hunk 4]
```diff
@@ -1,5 +1,5 @@
 import Responses from '../utils/response';
-import { getAllRequest, getOneRequest, getManagerRequest } from '../services/request.service';
+import { getAllRequest, getManagerRequest, getOneRequest } from '../services/request.service';
```
Comment (collinewait) [inline src/controllers/request.controller.js]: Is this change necessary?

File: src/controllers/roles.controller.js [hunk 5]
```diff
@@ -27,4 +27,5 @@ class Role {
     return Responses.handleSuccess(200, 'user role updated', res, userObject);
   }
 }
+
```
Comment (collinewait) [inline src/controllers/roles.controller.js]: Is this change necessary?

File: src/docs/api-specification.js [hunk 6]
```diff
@@ -8,7 +8,7 @@ const options = {
       title: 'Barefoot Nomad',
       version: '1.0.0',
       description:
-            'Making company travel and accomodation easy and convenient.',
+        'Making company travel and accomodation easy and convenient.',
```
Comment (collinewait) [inline src/docs/api-specification.js]: Is this change necessary?

File: src/middlewares/checkToken.js [hunk 7]
```diff
@@ -23,12 +23,12 @@ const decodeQueryToken = async (req, res, next) => {
...
-* @param {} res
-* @param {} next
-* @returns {undefined}
-*/
+ * function - decodes the token from the request
+ * @param {} req
+ * @param {} res
+ * @param {} next
+ * @returns {undefined}
+ */
```
Comment (collinewait) [inline src/middlewares/checkToken.js]: Are changes in this file necessary?

File: src/middlewares/roomsAvailability.js [hunk 8]
```diff
@@ -41,7 +42,8 @@ const checkForMultiCityRooms = async (req, res, next) => {
 
   const trips = req.body;
 
-  const rooms = trips.map((item) => item.rooms).flat();
+  const rooms = trips.map((item) => item.rooms)
+    .flat();
```
Comment (collinewait) [inline src/middlewares/roomsAvailability.js]: Are changes in this file necessary?

File: src/models/index.js [hunk 9]
```diff
@@ -17,11 +17,12 @@ readdirSync(__dirname)
...
-  if (db[modelName].associate) {
-    db[modelName].associate(db);
-  }
-});
+Object.keys(db)
+  .forEach((modelName) => {
+    if (db[modelName].associate) {
+      db[modelName].associate(db);
+    }
+  });
```
Comment (collinewait) [inline src/models/index.js]: Are changes in this file necessary?

File: src/routes/api/trips.route.js [hunk 10]
```diff
@@ -1,7 +1,7 @@
 import express from 'express';
 import trips from '../../controllers/trips.controller';
-import { validation, validateMultiCity } from '../../validation/validation';
-import { checkForRooms, checkForMultiCityRooms } from '../../middlewares/roomsAvailability';
+import { validateMultiCity, validation } from '../../validation/validation';
+import { checkForMultiCityRooms, checkForRooms } from '../../middlewares/roomsAvailability';
```
Comment (collinewait) [inline src/routes/api/trips.route.js]: Are changes in this file necessary?

File: src/routes/index.js [hunk 11]
```diff
@@ -2,12 +2,14 @@ import router from './api/index';
...
-      }, {})
-    });
+    return res.status(422)
+      .json({
+        errors: Object.keys(err.errors)
+          .reduce((errors, key) => {
+            errors[key] = err.errors[key].message;
+            return errors;
+          }, {})
+      });
```
Comment (collinewait) [inline src/routes/index.js]: Are changes in this file necessary?

File: src/scripts/migrate.js [hunk 12]
```diff
@@ -4,7 +4,10 @@ import { spawn } from 'child-process-promise';
 import config from '../config';
 import logger from '../utils/winston';
 
-const spawnOptions = { cwd: path.join(__dirname, '../..'), stdio: 'inherit' };
+const spawnOptions = {
+  cwd: path.join(__dirname, '../..'),
+  stdio: 'inherit'
+};
```
Comment (collinewait) [inline src/scripts/migrate.js]: Are changes in this file necessary?

File: src/scripts/rollback.js [hunk 13]
```diff
@@ -3,7 +3,10 @@ import { spawn } from 'child-process-promise';
 import config from '../config';
 import logger from '../utils/winston';
 
-const spawnOptions = { cwd: path.join(__dirname, '../..'), stdio: 'inherit' };
+const spawnOptions = {
+  cwd: path.join(__dirname, '../..'),
+  stdio: 'inherit'
+};
```
Comment (collinewait) [inline src/scripts/rollback.js]: Are changes in this file necessary?

File: src/services/Mailer.services.js [hunk 14]
```diff
@@ -5,8 +5,8 @@ import config from '../config';
 sgMail.setApiKey(config.SENDGRID_API_KEY);
 
 /**
-* Class representing a mailing services.
-*/
+ * Class representing a mailing services.
+ */
```
Comment (collinewait) [inline src/services/Mailer.services.js]: Are changes in this file necessary?

File: src/services/request.service.js [hunk 15]
```diff
@@ -4,7 +4,8 @@ import ErrorHandler from '../utils/error';
 const createRequest = async (userId, type) => {
   const request = await db.request.create(
     {
-      userId, type
+      userId,
+      type
```
Comment (collinewait) [inline src/services/request.service.js]: Are changes in this file necessary?

File: src/tests/RequestsApproval.test.js [hunk 16]
```diff
@@ -1,11 +1,7 @@
 import chaiHttp from 'chai-http';
-import { use, request, should } from 'chai';
+import { request, should, use } from 'chai';
 import app from '../app';
-import {
-  userfactory,
-  hotelfactory,
-  roomfactory
-} from './scripts/factories';
+import { hotelfactory, roomfactory, userfactory } from './scripts/factories';
```
Comment (collinewait) [inline src/tests/RequestsApproval.test.js]: Are changes in this file necessary?

File: src/tests/getRequests.test.js [hunk 17]
```diff
@@ -1,11 +1,7 @@
 import chaiHttp from 'chai-http';
-import { use, request, should } from 'chai';
+import { request, should, use } from 'chai';
 import app from '../app';
-import {
-  userfactory,
-  hotelfactory,
-  roomfactory
-} from './scripts/factories';
+import { hotelfactory, roomfactory, userfactory } from './scripts/factories';
```
Comment (collinewait) [inline src/tests/getRequests.test.js]: Are changes in this file necessary?

File: src/tests/roles.test.js [hunk 18]
```diff
@@ -1,6 +1,6 @@
 import chaiHttp from 'chai-http';
 // eslint-disable-next-line object-curly-newline
-import { use, request, expect, should } from 'chai';
+import { expect, request, should, use } from 'chai';
```
Comment (collinewait) [inline src/tests/roles.test.js]: Are changes in this file necessary?

File: src/tests/scripts/truncate.js [hunk 19]
```diff
@@ -8,6 +7,9 @@ import models from '../../models';
 export default async function truncate() {
   return Promise.all((Object.keys(models)).map((key) => {
     if (['sequelize', 'Sequelize'].includes(key)) return null;
-    return models[key].destroy({ where: {}, force: true });
+    return models[key].destroy({
+      where: {},
+      force: true
+    });
```
Comment (collinewait) [inline src/tests/scripts/truncate.js]: Are changes in this file necessary?

File: src/tests/trips.test.js [hunk 20]
```diff
@@ -1,11 +1,11 @@
 import chaiHttp from 'chai-http';
-import { use, request, should } from 'chai';
+import { request, should, use } from 'chai';
```
Comment (collinewait) [inline src/tests/trips.test.js]: Are changes in this file necessary?

File: src/tests/units/controllers/emailVerificationController.test.js [hunk 21]
```diff
@@ -1,4 +1,4 @@
-import chai, { expect, assert } from 'chai';
+import chai, { assert, expect } from 'chai';
```
Comment (collinewait) [inline src/tests/units/controllers/emailVerificationController.test.js]: Are changes in this file necessary?

File: src/tests/units/controllers/resetPassword.test.js [hunk 22]
```diff
@@ -1,4 +1,4 @@
-import chai, { expect, assert } from 'chai';
+import chai, { assert, expect } from 'chai';
```
Comment (collinewait) [inline src/tests/units/controllers/resetPassword.test.js]: Are changes in this file necessary?

File: src/tests/units/middlewares/checkToken.test.js [hunk 23]
```diff
@@ -1,4 +1,4 @@
-import chai, { expect, assert } from 'chai';
+import chai, { assert, expect } from 'chai';
```
Comment (collinewait) [inline src/tests/units/middlewares/checkToken.test.js]: Are changes in this file necessary?

File: src/tests/units/middlewares/validation.test.js [hunk 24]
```diff
@@ -1,9 +1,9 @@
-import chai, { expect, assert } from 'chai';
+import chai, { assert, expect } from 'chai';
```
Comment (collinewait) [inline src/tests/units/middlewares/validation.test.js]: Are changes in this file necessary?

File: src/tests/units/services/mailerServices.test.js [hunk 25]
```diff
@@ -9,8 +9,10 @@ describe('Mailing service', () => {
     const Mailing = new Mailer();
     Mailing.sendEmail(data.mailObject)
       .then((result) => {
-        expect(200).eql(result[0].statusCode);
+        expect(200)
+          .eql(result[0].statusCode);
         resolve();
-      }).then(done());
+      })
+      .then(done());
```
Comment (collinewait) [inline src/tests/units/services/mailerServices.test.js]: Are changes in this file necessary?

File: src/tests/userSignIn.test.js [hunk 26]
```diff
@@ -1,5 +1,5 @@
 import chaiHttp from 'chai-http';
-import { use, request, expect } from 'chai';
+import { expect, request, use } from 'chai';
```
Comment (collinewait) [inline src/tests/userSignIn.test.js]: Are changes in this file necessary?

File: src/tests/users.test.js [hunk 27]
```diff
@@ -1,5 +1,5 @@
 import chaiHttp from 'chai-http';
-import { use, request, expect } from 'chai';
+import { expect, request, use } from 'chai';
```
Comment (collinewait) [inline src/tests/users.test.js]: Are changes in this file necessary?

File: src/utils/error.js [hunk 28]
```diff
@@ -4,10 +4,10 @@
...
 export default class ErrorHandler extends Error {
   /**
-     * Create error handler.
-     * @param {string} message - The message to be passed for the error.
-     * @param {number} statusCode - The status code of the error.
-     */
+   * Create error handler.
+   * @param {string} message - The message to be passed for the error.
+   * @param {number} statusCode - The status code of the error.
+   */
```
Comment (collinewait) [inline src/utils/error.js]: Are changes in this file necessary?

File: src/utils/hash.js [hunk 29]
```diff
@@ -8,7 +8,7 @@ class Hash {
   /**
    *
    * @param {string} plainPassword
-   * @returns {strings} hashed password
+   * @returns {string} hashed password
```
Comment (collinewait) [inline src/utils/hash.js]: Is this change necessary?

File: src/utils/helper.js [hunk 30]
```diff
@@ -1,8 +1,9 @@
...
-   * @param {function} fn
-   * @returns {function} next
-   */
-const catchErrors = fn => (req, res, next) => fn(req, res, next).catch(next);
+ * use to wrap around controller to avoid try catch blocks
+ * @param {function} fn
+ * @returns {function} next
+ */
+const catchErrors = fn => (req, res, next) => fn(req, res, next)
+  .catch(next);
```
Comment (collinewait) [inline src/utils/helper.js]: Are changes in this file necessary?

File: src/utils/helpers.js [hunk 31]
```diff
@@ -1,5 +1,10 @@
 // eslint-disable-next-line no-unused-vars
-const randomPw = len => Array(len).fill(0).map(() => Math.random().toString(36).charAt(2)).join('');
+const randomPw = len => Array(len)
+  .fill(0)
+  .map(() => Math.random()
+    .toString(36)
+    .charAt(2))
+  .join('');
```
Comment (collinewait) [inline src/utils/helpers.js]: Are changes in this file necessary?

File: src/utils/schemas.js [hunk 32]
```diff
@@ -1,85 +1,162 @@
...
-  email: Joi.string().strict().trim().email()
-    .required(),
-  password: Joi.string().alphanum().min(8).required()
-}).options({
-  abortEarly: false,
-  language: {
-    key: '{{key}} '
-  }
-});
+const signupSchema = Joi.object()
```
Comment (collinewait) [inline src/utils/schemas.js]: Are changes in this file necessary?

File: src/validation/validation.js [hunk 33]
```diff
@@ -1,6 +1,7 @@
 import Joi from '@hapi/joi';
 import Responses from '../utils/response';
 import Schemas from '../utils/schemas';
+
```
Comment (collinewait) [inline src/validation/validation.js]: Are changes in this file necessary?

**Review by nshaphan — DISMISSED** [review]
Great work @gildniy, However, I don't know if it is you or your prettier configuration, the way you are formatting code is reducing code readability and making it harder for people to review changes.

**Comment by nshaphan** [comment]
Fix conflicts, I believe that your feature is well implemented and ready to be merged

**Comment by gildniy** [comment]
> Great work @gildniy, However, I don't know if it is you or your prettier configuration, the way you are formatting code is reducing code readability and making it harder for people to review changes.

It's the IDE I use. It tries to list different element alphabetically and reformatting some lines ... Let me undo them.

**Review by nshaphan — APPROVED** [review]
Keep the good work @gildniy now it's working and looking great

**Review by RaymondMwaura — APPROVED** [review]
Great work Gildas. You have acted on all the feedback. Keep it up.

**Review by collinewait — APPROVED** [review]
